stdlib/public/core/Substring.swift.gyb

```diff
@@ -19,7 +19,17 @@ extension String {
   ///
   /// - Complexity: O(*n*), where *n* is the length of `substring`.
   public init(_ substring: Substring) {
-    self = String(substring._slice)
+    let x = substring._wholeString
+    let start = substring.startIndex
+    let end = substring.endIndex
+    let u16 = x._core[start.encodedOffset..<end.encodedOffset]
+    if start.samePosition(in: x.unicodeScalars) != nil
+    && end.samePosition(in: x.unicodeScalars) != nil {
+      self = String(_StringCore(u16))
```

Comment:
AFAICT it calls the one from the `RangeReplaceableCollection` conformance, which copies.

Yes, the `samePosition` checks looks for unpaired surrogates.

In Swift 3, `String` never contains invalid UTF-16 except when is holding an `NSString`, and although that's not where we want to be in the long run I was not prepared to change that in this PR.  It was so hard to get all the tests to pass (4 days, including nights and weekends) that I may have been over-cautious in this part of the code.  Feel free to play with it if you have time.